compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -620,6 +620,55 @@ impl<'a> Parser<'a> {
             );
         }
 
+        // Possible missing `struct` or `enum` keyword
+        if let TokenKind::Ident(_, _) = self.prev_token.kind
+            && let TokenKind::OpenDelim(Delimiter::Brace) = self.token.kind
+            && let [TokenType::Token(token::Not), TokenType::Token(token::PathSep)] = expected[..]
+        {
+            let mut lookahead = self.clone();
+            for position in 1..1000 {
+                lookahead.bump();
+                if lookahead.token == token::CloseDelim(Delimiter::Brace)
+                    || lookahead.token == token::Colon
+                {
+                    let (article, kw) = if lookahead.token == token::Colon || position == 1 {
+                        ("a", "struct")
+                    } else {
+                        ("an", "enum")
+                    };
+                    err.span_suggestion(
+                        self.prev_token.span.shrink_to_lo(),
+                        format!("if this is {article} {kw} definition, use the `{kw}` keyword"),
+                        format!("{kw} "),
+                        Applicability::MaybeIncorrect,
+                    );
+                    break;
+                }
+            }
+        }
+
+        // Possible missing `fn` keyword
+        if let TokenKind::Ident(_, _) = self.prev_token.kind
+            && let TokenKind::OpenDelim(Delimiter::Parenthesis) = self.token.kind
+        {
+            let mut lookahead = self.clone();
+            for _ in 1..100 {
+                lookahead.bump();
+                if lookahead.token == token::Semi {
+                    break;
+                }
+                if lookahead.token == token::OpenDelim(Delimiter::Brace) {
```

Comment:
I'd suggest look inside the parameters. If there are no parameters, e.g. `foo() {` / `foo();` you are expected to find a semicolon with only two bumps.

If there are, look for colons. If one writes `foo(x: u32)` we're pretty sure they are writing a function. If we find no colon before the closing parenthesis or before a comma `foo(x, y)` or `foo(x)`, then that looks like a function call. This can be done with up to two bumps since the opening parenthesis as well.